src/compute/src/compute_state.rs

```diff
@@ -415,26 +410,33 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
     /// Send progress information to the coordinator.
     pub fn report_compute_frontiers(&mut self) {
         let mut new_uppers = Vec::new();
+        let mut new_frontier = Antichain::new();
 
-        let mut update_frontier = |id, new_frontier: &Antichain<Timestamp>| {
-            let reported_frontier = self
-                .compute_state
-                .reported_frontiers
-                .get_mut(&id)
-                .unwrap_or_else(|| panic!("Frontier update for untracked identifier: {id}"));
+        for (&id, collection) in self.compute_state.collections.iter_mut() {
+            if let Some(traces) = self.compute_state.traces.get_mut(&id) {
+                traces.oks_mut().read_upper(&mut new_frontier);
+            } else if let Some(frontier) = &collection.sink_write_frontier {
+                new_frontier.clone_from(&frontier.borrow());
```

Comment:
Is it a documented invariant somewhere that we cannot have `traces` for a sink? The previous version of the code was not strict in this same sense. So perhaps it is good to either document this better or revisit the above.

Reply:
Yes, a collection can only be either an index, a persist sink, or a subscribe. The previous code also assumed that implicitly, as I'm pretty sure the frontier reporting logic would have been non-sensical if you had a collection with multiple frontiers (I think it would have reported the larger of the uppers?).

I'll add a comment and an assert so we immediately notice should that assumption ever change.